lib/config.js

```diff
@@ -352,11 +352,17 @@ const CONFIG_SYNTAX_HELP = '  module.exports = function(config) {\n' +
   '  };\n'
 
 function parseConfig (configFilePath, cliOptions, parseOptions) {
+  const promiseConfig = parseOptions && parseOptions.promiseConfig === true
+  const throwErrors = parseOptions && parseOptions.throwErrors === true
   function fail () {
     log.error(...arguments)
-    if (parseOptions && parseOptions.throwErrors === true) {
+    if (throwErrors === true || promiseConfig === true) {
```

Comment:
IMO it is cleaner to drop `|| promiseConfig === true` part, so flags are independent and one does not imply the other.

Reply:
That's fine. Throwing vs rejecting is still mutually exclusive, but it can easily be worked with.

My instinct is to avoid writing the same error twice.  do I self contain the thing? or do I avoid maintaining the same thing in multiple locations? Here is what I'm thinking:

```js
  function fail () {
    log.error(...arguments)
    const errorMessage = Array.from(arguments).join(' ')
    const err = new Error(errorMessage)
    if (promiseConfig === true) {
      return Promise.reject(err)
    } else if (throwErrors === true) {
      throw err
    } else {
```

Reply:
I think the logic should be like this:

```
if (throwErrors) {
  if (promiseConfig) return Promise.reject(err)
  else throw err
} else {
  process.exit(1)
}
```

E.g. promiseConfig == true should not imply throwErrors == true (which controls process.exit behavior). Does it make sense?

Reply:
I am ok with using that logic, since it accomplishes the same thing. The difference in the end result is requiring `throwErrors` to be set in order to use promises.

> promiseConfig == true should not imply throwErrors == true

I agree. I was trying to go for that but my thinking was following a different path.

<details>

Throwing and promising are both things that should disable `process.exit`.

Promises work on fulfillment and rejection. If we don't reject with errors, then we are breaking the expectation the consumer has of using promises.

I guess you could say I am concerned about scenarios where `{throwsErrors: false /* or undefined */, promiseConfig: true}` occur. Instead of a rejection, the process could exit on failure. When using promises, I would not expect that kind of behavior. Since promising configs is a new feature, it seems intuitive to me to avoid a `process.exit` entirely with it.

I might be getting hung up on the choice of words here. `throwsErrrors` vs a hypothetical `dontExit`, which I probably wouldn't be as worried about as it would be `throwsErrors` with a different name.

Since `throwsErrors` is logic that will be cleaned up in a future major release, I'm ok with using the code you suggested above. It absolutely meets the internal and public API goals of this PR.

---

ASIDE: I can see `promiseConfig` staying in the future/cleaned up code (in some form) so that public API users can chose between sync and async. Users would have to accept that sync won't support async configs. (Not sure how I feel about that.) An alternative would be to offer 2 functions (sync and async) and remove the option entirely from the public API.

```js
function parseConfigBase(configFilePath, cliOptions, parseOptions, promiseConfig) {
  // The same function we have now, only not exported
  // Including code review changes
}

// This would be the sync function for backwards compatibility
// otherwise I would have suggested that this be `parseConfigSync` or `parseConfig.sync`
function parseConfig(configFilePath, cliOptions, parseOptions) {
  return parseConfigBase(configFilePath, cliOptions, parseOptions, false)
}

// This name makes it clear that async uses promises instead of the callback pattern
parseConfig.promise = function promiseParsedConfig(configFilePath, cliOptions, parseOptions) {
  return parseConfigBase(configFilePath, cliOptions, parseOptions, true)
}
```

This aside is more thinking aloud for the future and should not affect the direction taken by this PR.

</details>

TL;DR : I'll make the change you suggested.

Reply:
Thanks for detailing it out. The alternative design makes total sense as well.

However, I expect both of these flags to be removed in the future for cleaner code and API simplicity over a hypothetical use case and that's why I suggested this change.